### DISCUSSION TITLE
Packaging for NixOS

### DISCUSSION THREAD

**Discussion by Jaculabilis** [alangrainger/immich-public-proxy#34] [Feature Requests]
Thanks for making this! I run NixOS on my media server, and the latest NixOS release includes an Immich module that I'm going to use. I saw IPP on /r/selfhosted and wanted to use it, so I figured out how to package it for NixOS. Before contributing it to [nixpkgs](https://github.com/NixOS/nixpkgs), I thought it would be polite (1) to check in with you, the author, in case you'd prefer IPP not to be added to nixpkgs; and (2) to offer to upstream some of the patches I had to apply.

The nixpkgs branch that I would PR is [here](https://github.com/NixOS/nixpkgs/compare/master...Jaculabilis:nixpkgs:immich-public-proxy), if you're curious. It includes some patches that would be applied, which are derived from [my forked IPP repo](https://github.com/alangrainger/immich-public-proxy/compare/main...Jaculabilis:immich-public-proxy:main). In particular, the three changes are:

* [Add a `bin` property to `package.json`](https://github.com/alangrainger/immich-public-proxy/commit/c577f101e3e3f6ade0ad06b5d344e228a8f68b0f) ([npm docs](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#bin)). I know that in nixpkgs, this is needed to get a `bin/immich-public-proxy` executable; I'm not sure it it would do anything in the build workflow that you have.
* [Allow the port to be set by `IPP_PORT` envvar instead of always 3000](https://github.com/alangrainger/immich-public-proxy/commit/a377081c80dab13320c4511630dc430c1208a49f). In Docker this doesn't matter since you can remap the port, but the NixOS module runs it outside a container, so it needs a different way to change the port.
* Add `app/package-lock.json`. Nix requires a lockfile to do a deterministic build. I noted that you added `package-lock.json` to the `.gitignore` in https://github.com/alangrainger/immich-public-proxy/commit/690bd3f688c10211ca5f0ef5f8f760253d317f76, which I assume is because you are using Docker images to provide the same build on different machines. Personally, I would recommend having a lockfile even with Docker, but if your workflow requires it be absent, I can work around that.

If you're open to it, I'll create PRs for these changes and adjust my Nix packaging for the updated upstream.

**Comment by alangrainger** [comment]
Sounds great! Please do submit the PR.

The only one I'm not certain about is the `bin` property, simply because I don't know what (if any) effect it might have on the build. But I can easily test that.

**Reply by Jaculabilis** [reply]
Based on the npm docs, it only matters when the package is installed globally or as a dependency.

> When this package is installed globally, that file will be either linked inside the global bins directory or a cmd (Windows Command File) will be created which executes the specified file in the `bin` field, so it is available to run by `name` or `name.cmd` (on Windows PowerShell). When this package is installed as a dependency in another package, the file will be linked where it will be available to that package either directly by npm exec or by name in other scripts when invoking them via npm run-script.

IPP in Docker is built and passed to a Node server rather than being installed globally, and it isn't uploaded to npm so it can't even be a dependency of another package. So for all current purposes, `bin` should effectively be a no-op.

**Comment by Jaculabilis** [comment]
Opened #35, #36, and #37. When those are resolved and there's a new version, I'll update the Nix packaging.

**Comment by Jaculabilis** [comment]
NixOS packaging review is underway at https://github.com/NixOS/nixpkgs/pull/362907. Opened #40 for one more patch.

**Reply by alangrainger** [reply]
Merged, thanks. That's quite exciting!

Do you need me to make a new release incorporating those changes, or is it fine with them just in the main branch?

**Reply by Jaculabilis** [reply]
A new release would be good. Let me confirm with the reviewer first that there aren't any other changes that could result in patches.

**Reply by Jaculabilis** [reply]
The reviewer hasn't gotten back to me yet, which might be because of the holidays. I'm going to optimistically assume I won't need any more patches to IPP and any remaining changes will be in the Nix code, so feel free to cut a new release at your leisure.

**Reply by alangrainger** [reply]
v.1.5.4 released to incorporate the above PRs.

**Reply by Jaculabilis** [reply]
Thank you! I'll update the packaging PR as time permits around holiday events.

**Comment by Jaculabilis** [comment]
The package and module are merged! It should be available in `nixpkgs-unstable` now, and in the 25.05 release at some point. Thanks for your help!
